File: ui/src/features/dashboard/kernel-error.ts
/** Kernel start / config-apply stable error codes for ops diagnostics. */

export type KernelErrorCode =
  | "config_invalid"
  | "config_missing"
  | "restart_failed"
  | "start_failed"
  | "permission"
  | "unknown"

const HINT_KEYS: Record<string, string> = {
  config_invalid: "dashboard.errorHintConfigInvalid",
  config_missing: "dashboard.errorHintConfigMissing",
  restart_failed: "dashboard.errorHintRestartFailed",
  start_failed: "dashboard.errorHintStartFailed",
  permission: "dashboard.errorHintPermission",
  unknown: "dashboard.errorHintUnknown",
}

export function kernelErrorHintKey(code?: string): string {
  if (!code) return HINT_KEYS.unknown
  return HINT_KEYS[code] ?? HINT_KEYS.unknown
}

export function classifyKernelErrorMessage(message?: string): KernelErrorCode {
  const lower = (message ?? "").toLowerCase()
  if (!lower) return "unknown"
  if (lower.includes("no such file") || lower.includes("not exist") || lower.includes("cannot find")) {
    return "config_missing"
  }
  if (lower.includes("permission denied") || lower.includes("operation not permitted")) {
    return "permission"
  }
  if (lower.includes("restart failed")) return "restart_failed"
  if (
    lower.includes("decode")
    || lower.includes("invalid")
    || lower.includes("unmarshal")
    || lower.includes("unknown field")
    || lower.includes("missing required")
    || lower.includes("legacy")
    || lower.includes("required")
  ) {
    return "config_invalid"
  }
  if (
    lower.includes("start failed")
    || lower.includes("listen")
    || lower.includes("bind")
    || lower.includes("address already in use")
    || lower.includes("factory failed")
  ) {
    return "start_failed"
  }
  return "unknown"
}

export function resolveKernelErrorCode(item: {
  error?: string
  error_code?: string
  last_error?: string
  last_error_code?: string
}): string | undefined {
  const code = item.error_code?.trim() || item.last_error_code?.trim()
  const message = item.error?.trim() || item.last_error?.trim()
  if (!message && !code) return undefined
  return code || classifyKernelErrorMessage(message)
}

export function kernelLastErrorClipboardText(status: {
  running?: boolean
  config_path?: string
  version?: string
  last_error?: string
  last_error_code?: string
  last_error_at?: string
}): string {
  const error = status.last_error?.trim()
  if (!error) return ""
  const code = resolveKernelErrorCode(status)
  const lines = [
    status.running === undefined ? "" : `running: ${status.running ? "true" : "false"}`,
    status.config_path?.trim() ? `config: ${status.config_path.trim()}` : "",
    status.version?.trim() ? `version: ${status.version.trim()}` : "",
    code ? `code: ${code}` : "",
    `error: ${error}`,
    status.last_error_at?.trim() ? `at: ${status.last_error_at.trim()}` : "",
  ].filter(Boolean)
  return lines.join("\n")
}

File: internal/core/kernel_error.go
package core

import (
	"errors"
	"net/netip"
	"os"
	"strings"
)

// 内核启动 / 配置应用稳定错误码，供前端展示可操作提示。
const (
	KernelErrorConfigInvalid   = "config_invalid"
	KernelErrorConfigMissing   = "config_missing"
	KernelErrorRestartFailed   = "restart_failed"
	KernelErrorStartFailed     = "start_failed"
	KernelErrorPermission      = "permission"
	KernelErrorIPv6Unavailable = "ipv6_unavailable"
	KernelErrorUnknown         = "unknown"
)

// ClassifyKernelError 将内核/配置应用失败映射为稳定错误码。
func ClassifyKernelError(msg string, err error) string {
	if isIPv6AddressSetupError(msg) {
		return KernelErrorIPv6Unavailable
	}
	if err != nil && isIPv6AddressSetupError(err.Error()) {
		return KernelErrorIPv6Unavailable
	}
	if code := classifyKernelErrorValue(err); code != "" {
		return code
	}
	return classifyKernelErrorMessage(msg)
}

func isIPv6AddressSetupError(msg string) bool {
	lower := strings.ToLower(msg)
	if strings.Contains(lower, "set ipv6 address") {
		return true
	}
	for _, detail := range strings.Split(lower, "add address ")[1:] {
		fields := strings.Fields(detail)
		if len(fields) == 0 {
			continue
		}
		prefix, err := netip.ParsePrefix(strings.TrimSuffix(fields[0], ":"))
		if err == nil && prefix.Addr().Is6() {
			return true
		}
	}
	return false
}

func classifyKernelErrorValue(err error) string {
	if err == nil {
		return ""
	}
	if errors.Is(err, os.ErrNotExist) {
		return KernelErrorConfigMissing
	}
	if errors.Is(err, os.ErrPermission) {
		return KernelErrorPermission
	}
	return ""
}

func classifyKernelErrorMessage(msg string) string {
	lower := strings.ToLower(strings.TrimSpace(msg))
	if lower == "" {
		return KernelErrorUnknown
	}
	switch {
	case strings.Contains(lower, "no such file"), strings.Contains(lower, "not exist"), strings.Contains(lower, "cannot find"):
		return KernelErrorConfigMissing
	case strings.Contains(lower, "permission denied"), strings.Contains(lower, "operation not permitted"):
		return KernelErrorPermission
	case strings.Contains(lower, "restart failed"):
		return KernelErrorRestartFailed
	case strings.Contains(lower, "decode"),
		strings.Contains(lower, "invalid"),
		strings.Contains(lower, "unmarshal"),
		strings.Contains(lower, "unknown field"),
		strings.Contains(lower, "missing required"),
		strings.Contains(lower, "legacy"),
		strings.Contains(lower, "missing "),
		strings.Contains(lower, "required"):
		return KernelErrorConfigInvalid
	case strings.Contains(lower, "start failed"),
		strings.Contains(lower, "listen"),
		strings.Contains(lower, "bind"),
		strings.Contains(lower, "address already in use"),
		strings.Contains(lower, "factory failed"):
		return KernelErrorStartFailed
	default:
		return KernelErrorUnknown
	}
}
